### PR TITLE
MicrosoftGraphListener - Add self deployed microsoft graph listener

### PR DESCRIPTION
**Pull request by abaumgarten**
## Status
- [] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/28041

## Description
Fixed an issue where self deployed option was not supported.

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md [hunk 1]
```diff
@@ -1,2 +1,24 @@
...
+
+Follow these steps for a self-deployed configuration.
+
+1. To use a self-configured Azure application, you need to add a new MS Graph App Registration in the Azure Portal. To add the registration, refer to the following [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
+2. Make sure the following permissions are granted for the app registration:
+   -  API/Permission name `openid` of type `Delegated`
+   -  API/Permission name `profile` of type `Delegated`
+   -  API/Permission name `Mail.ReadWrite` of type `Delegated`
+   -  API/Permission name `Mail.Send` of type `Delegated`
+   -  API/Permission name `Mail.Read` of type `Delegated`
```
Comment (ronykoz) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md:14]: did you make sure that these are the minimal permission set?

**Review by ronykoz — APPROVED** [review]
Looks great!

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md (hunk 1 above)
Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md:14]: I think that these are the permissions we have to include. 
@Itay4  am I right?

**Review by DeanArbel — APPROVED** [review]

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 2]
```diff
@@ -320,12 +320,12 @@ def _get_self_deployed_token_auth_code(
         Returns:
             tuple: An access token, its expiry and refresh token.
         """
-        data = {
-            'client_id': self.client_id,
-            'client_secret': self.client_secret,
-            'resource': self.resource if not resource else resource,
-            'redirect_uri': self.redirect_uri
-        }
+        data = assign_params(
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: why is this change needed?
this could lead to unexpected behavior in the other microsoft integrations

File: Packs/MicrosoftGraphListener/ReleaseNotes/1_0_8.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Microsoft Graph Mail Single User
+- Fixed an issue where **self deployed** option was not supported. 
+- Upgraded the Docker image to demisto/crypto:1.0.0.12979.- Upgraded the Docker image to demisto/crypto:1.0.0.12979.
```
Comment (Itay4) [inline Packs/MicrosoftGraphListener/ReleaseNotes/1_0_8.md:5]: fix release notes here

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py (hunk 2 above)
Comment (DeanArbel) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: I suggested this change since leaving the resource key with an empty value resulted in the following 400 error:
image

Comment (Itay4) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: how come this did not come up in the past in other integrations?

File: Packs/MicrosoftGraphListener/ReleaseNotes/1_0_8.md (hunk 3 above)
Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/ReleaseNotes/1_0_8.md]: Done

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md (hunk 1 above)
Comment (Itay4) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md:14]: these shouldnt be in here, but in the integration doc
also we already have this documented in https://xsoar.pan.dev/docs/reference/articles/microsoft-integrations---authentication so you should just refer to that article
see how its done in other microsoft integrations

Comment (Itay4) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md:14]:
```suggestion
   -  API/Permission name `User.Read` of type `Delegated`
```

**Comment by Itay4** [comment]
please add the new parameters to the integration doc

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml [hunk 4]
```diff
@@ -15,6 +15,14 @@ configuration:
   name: enc_key
   required: true
   type: 4
+- display: Authorization code (received from the authorization step - see Detailed Instructions (?) section)
```
Comment (Itay4) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml:18]:
```suggestion
- display: Authorization code (required for self-deployed Azure app)
```

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml [hunk 5]
```diff
@@ -15,6 +15,14 @@ configuration:
...
+  name: auth_code
+  required: false
+  type: 4
+- display: Application redirect URI (for self-deployed mode)
```
Comment (Itay4) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml:22]:
```suggestion
- display: Application redirect URI (required for self-deployed Azure app)
```

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml [hunk 6]
```diff
@@ -15,6 +15,14 @@ configuration:
   name: enc_key
```
Comment (Itay4) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml:15]: for the id, token and key, update the display names to be the same as in Graph Mail

Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml:15]: Done

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml (hunk 4 above)
Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml]: Done

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml (hunk 5 above)
Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener.yml]: Done

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md (hunk 1 above)
Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md]: Ok, copied the description from MicrosoftGraphMail

Comment (abaumgarten) [reply Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md]: This is not appear anymore

**Comment by abaumgarten** [comment]
> please add the new parameters to the integration doc

Added.

**Comment by Itay4** [comment]
@abaumgarten did we test tokens are refreshed as expected after expiration? 

also note [my comment](https://github.com/demisto/content/pull/9799#discussion_r517187258), would like to understand better why this change is required

**Comment by abaumgarten** [comment]
> @abaumgarten did we test tokens are refreshed as expected after expiration?
> Yes, I can still using the integration's commands with the same configuration from more than a hour ago.

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py (hunk 2 above)
Comment (DeanArbel) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: I'm not sure why this didn't come up before, but it seems that `resource` [was replaced](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/azure-ad-endpoint-comparison#scopes-not-resources) with `scope` in `v2`.

Comment (Itay4) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: did we test this against the integrations that use auth code flow?

File: Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md [hunk 7]
```diff
@@ -1,2 +1,24 @@
...
+- ***msgraph-mail-create-draft***
+- ***msgraph-mail-send-draft***
+- ***msgraph-mail-reply-to***
+- ***send-mail***
+
+To use these commands and to fetch incidents,
+you will need to add to your application the *Mail.Send application* permission (not delegated),
+and re-authorize your integration's instance.
+
+If you do not wish to use these commands, you may keep your integration credentials the same.
```
Comment (kirbles19) [inline Packs/MicrosoftGraphListener/Integrations/MicrosoftGraphListener/MicrosoftGraphListener_description.md:24]: If you don't need to use these commands you can retain your current integration credentials.

**Comment by kirbles19** [comment]
@abaumgarten Done.

**Review by guykeller — APPROVED** [review]

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py (hunk 2 above)
Comment (abaumgarten) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:323]: I have just tested all those integrations that use auth code flow and seemed to be working as expected
